Homework 3
Authors: Daniil Morozkov,

Question 1
Describe your data and take reasonable steps to clean the data. 

In [288]:
# import libraries
import pandas as pd
import numpy as np

In [289]:
# upload dataset
crsp = pd.read_excel('CRSP_Dividends.xlsx')

In [290]:
# check info
crsp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 458220 entries, 0 to 458219
Data columns (total 15 columns):
 #   Column   Non-Null Count   Dtype         
---  ------   --------------   -----         
 0   PERMNO   458220 non-null  int64         
 1   date     458220 non-null  datetime64[ns]
 2   SHRCD    458220 non-null  int64         
 3   PERMCO   458220 non-null  int64         
 4   DISTCD   458220 non-null  int64         
 5   DIVAMT   458220 non-null  float64       
 6   PRC      458220 non-null  float64       
 7   VOL      416627 non-null  float64       
 8   RET      458220 non-null  float64       
 9   BID      227403 non-null  float64       
 10  ASK      227403 non-null  float64       
 11  SHROUT   458220 non-null  int64         
 12  OPENPRC  196763 non-null  float64       
 13  RETX     458220 non-null  float64       
 14  vwretd   458220 non-null  float64       
dtypes: datetime64[ns](1), float64(9), int64(5)
memory usage: 52.4 MB


In [291]:
# show few entries
crsp.head()

PERMNO       date  SHRCD  PERMCO  DISTCD  DIVAMT     PRC     VOL       RET  \
0   10001 1986-03-10     11    7953    1232   0.095 -6.2500   100.0  0.015200   
1   10001 1986-06-09     11    7953    1232   0.105 -6.1875  1050.0  0.016970   
2   10001 1986-09-08     11    7953    1232   0.105  6.7500  1610.0  0.054615   
3   10001 1986-12-08     11    7953    1232   0.105  6.5000   400.0  0.016154   
4   10001 1987-03-09     11    7953    1232   0.105  6.1250   650.0  0.027629   

     BID   ASK  SHROUT  OPENPRC      RETX    vwretd  
0    NaN   NaN     985      NaN  0.000000  0.004365  
1    NaN   NaN     985      NaN  0.000000 -0.019374  
2  6.375  6.75     985      NaN  0.038462 -0.010801  
3  6.500  7.00     991      NaN  0.000000 -0.000944  
4  5.875  6.25     991      NaN  0.010309 -0.006488

In [292]:
# check if there are any dividend amounts less than 0.01
(crsp['DIVAMT']<0.01).any()

True

In [293]:
# remove dividend amounts less than 0.01
crsp = crsp.loc[crsp['DIVAMT'] >= 0.01]

In [294]:
# Negative prices indicate a stock that did not trade that day; common practice is to set those equal to the absolute value of the price.
crsp['PRC'] = abs(crsp['PRC'])

In [295]:
# check if there are any stocks that have a market capitalization below $50M
crsp['MKT'] = crsp['SHROUT']*crsp['PRC']
(crsp['MKT']<50000).any()

True

In [296]:
# remove entries that have a market capitalization below $50M
crsp = crsp.loc[crsp['SHROUT']*crsp['PRC'] >= 50000]

In [297]:
# check again, make sure everything worked out
(crsp['MKT']<50000).any()

False

In [298]:
# •	Importantly, read through the remainder of this assignment and think about how you should handle observations where VOL = 0. Let's remove them.
crsp = crsp.loc[crsp['VOL']>0]

In [299]:
# show the results of data cleaning
crsp.describe()

PERMNO                           date          SHRCD  \
count  322567.000000                         322567  322567.000000   
mean    47261.920472  1995-01-14 20:40:07.760248192      10.871881   
min     10001.000000            1962-01-02 00:00:00      10.000000   
25%     22955.000000            1982-11-12 00:00:00      11.000000   
50%     44329.000000            1995-06-14 00:00:00      11.000000   
75%     75333.000000            2008-05-28 00:00:00      11.000000   
max     93429.000000            2023-06-30 00:00:00      11.000000   
std     26574.565424                            NaN       0.334223   

              PERMCO         DISTCD         DIVAMT           PRC  \
count  322567.000000  322567.000000  322567.000000  322567.00000   
mean    19186.866784    1232.696131       0.253194      37.29405   
min         4.000000    1202.000000       0.010000       1.01000   
25%      9449.000000    1232.000000       0.100000      18.71000   
50%     20834.000000    1232.000000       0.190000      28.20000   
75%     22449.000000    1232.000000       0.320000      42.87500   
max     59543.000000    1294.000000      85.000000    4280.04004   
std     12216.603207       5.719189       0.501875      54.72760   

                VOL            RET            BID            ASK  \
count  3.225670e+05  322567.000000  192594.000000  192594.000000   
mean   6.705685e+05       0.002409      39.678352      39.952865   
min    1.000000e+00      -0.729864       1.000000       1.010000   
25%    5.725000e+03      -0.008268      17.800000      18.070000   
50%    3.500000e+04       0.002105      27.850000      28.125000   
75%    2.680610e+05       0.012213      44.550000      44.830000   
max    4.634910e+08       0.696822    4261.959960    4280.040040   
std    3.752105e+06       0.022643      67.430465      67.679871   

             SHROUT        OPENPRC           RETX         vwretd           MKT  
count  3.225670e+05  181549.000000  322567.000000  322567.000000  3.225670e+05  
mean   1.054957e+05      40.949164      -0.005888       0.000430  5.238313e+06  
min    1.910000e+02       0.500000      -0.878553      -0.171346  5.000000e+04  
25%    7.724000e+03      18.200000      -0.017045      -0.003954  1.671280e+05  
50%    2.016200e+04      28.750000      -0.005435       0.000806  5.288400e+05  
75%    6.124550e+04      46.062500       0.005025       0.005242  2.112365e+06  
max    1.700180e+07    4411.959960       0.691589       0.114921  2.813308e+09  
std    4.347917e+05      69.597442       0.025614       0.010174  2.992906e+07

Question 2
For each observation, calculate the closing price as of the prior day. You should be able to do this using the current PRC and the return. (Think about which return to use.) Create a column of (Pt-1 – Pt) / D. Describe the mean and range of this variable. In a markdown cell, discuss the findings – what does this variable signify? Is there a potential investment opportunity?

In [300]:
# reformat the original return formula to solve for Pt-1
crsp['PrPRC'] = crsp['PRC']/(crsp['RETX']+1)

In [301]:
# create a column of (Pt-1 – Pt) / D.
crsp['DIVDELTA'] = (crsp['PrPRC'] - crsp['PRC']) / crsp['DIVAMT']

In [302]:
# describe the mean and range of this variable
crsp['DIVDELTA'].describe()

count    322567.000000
mean          0.658522
std          10.621310
min        -906.248481
25%          -0.833365
50%           0.736883
75%           2.272746
max        1071.995035
Name: DIVDELTA, dtype: float64

In a markdown cell, discuss the findings – what does this variable signify? Is there a potential investment opportunity?

Tha variable describes relative change in close price before the ex-dividend day and at the ex-dividend day.
It might seem as a good investing opportunity, because on average (mean,0.658522) the price increases around the ex-dividend day, however by looking at large range min,-906.248481 : max,1071.995035 and large std = 10.621310, the strategy becomes increasingly risky, so sometimes we may even lose all of our funds from such price swings.


In [303]:
# upload t bill dataset
irx = pd.read_csv('^IRX.csv',  parse_dates=[0], index_col=[0])
irx['Tbill yield filled'] = irx['Close'].fillna(method='ffill')

In [304]:
# compute Nt by checking difference between dates
irx['Nt'] = irx.index.to_series().diff().dt.days.fillna(1)
print(irx['Nt'])

Date
1960-01-04    1.0
1960-01-05    1.0
1960-01-06    1.0
1960-01-07    1.0
1960-01-08    1.0
             ... 
2022-07-21    1.0
2022-07-22    1.0
2022-07-24    2.0
2022-07-25    1.0
2022-07-26    1.0
Name: Nt, Length: 19450, dtype: float64


In [305]:
# find risk free rate
irx['rf'] = irx['Tbill yield filled'] * irx['Nt']/36500
irx['rf']

Date
1960-01-04    0.000124
1960-01-05    0.000125
1960-01-06    0.000128
1960-01-07    0.000127
1960-01-08    0.000126
                ...   
2022-07-21    0.000065
2022-07-22    0.000065
2022-07-24    0.000129
2022-07-25    0.000066
2022-07-26    0.000068
Name: rf, Length: 19450, dtype: float64

In [306]:
# filter irx based on date prior to merge
start_date = crsp['date'].iloc[0]
end_date = crsp['date'].iloc[-1]
irx = irx[(irx.index >= start_date) & (irx.index <= end_date)]

In [308]:
# merge two datasets
crsp = pd.merge(crsp, irx, left_on='date', right_on=irx.index, how='left')

In [307]:
# calculate daily returns
crsp['DailyRets'] = (crsp['PRC'] - crsp['PrPRC']) / crsp['PrPRC']

In [309]:
# check if all looks good
crsp.describe()

PERMNO                           date          SHRCD  \
count  322567.000000                         322567  322567.000000   
mean    47261.920472  1995-01-14 20:40:07.760248192      10.871881   
min     10001.000000            1962-01-02 00:00:00      10.000000   
25%     22955.000000            1982-11-12 00:00:00      11.000000   
50%     44329.000000            1995-06-14 00:00:00      11.000000   
75%     75333.000000            2008-05-28 00:00:00      11.000000   
max     93429.000000            2023-06-30 00:00:00      11.000000   
std     26574.565424                            NaN       0.334223   

              PERMCO         DISTCD         DIVAMT           PRC  \
count  322567.000000  322567.000000  322567.000000  322567.00000   
mean    19186.866784    1232.696131       0.253194      37.29405   
min         4.000000    1202.000000       0.010000       1.01000   
25%      9449.000000    1232.000000       0.100000      18.71000   
50%     20834.000000    1232.000000       0.190000      28.20000   
75%     22449.000000    1232.000000       0.320000      42.87500   
max     59543.000000    1294.000000      85.000000    4280.04004   
std     12216.603207       5.719189       0.501875      54.72760   

                VOL            RET            BID  ...      DailyRets  \
count  3.225670e+05  322567.000000  192594.000000  ...  322567.000000   
mean   6.705685e+05       0.002409      39.678352  ...      -0.005888   
min    1.000000e+00      -0.729864       1.000000  ...      -0.878553   
25%    5.725000e+03      -0.008268      17.800000  ...      -0.017045   
50%    3.500000e+04       0.002105      27.850000  ...      -0.005435   
75%    2.680610e+05       0.012213      44.550000  ...       0.005025   
max    4.634910e+08       0.696822    4261.959960  ...       0.691589   
std    3.752105e+06       0.022643      67.430465  ...       0.025614   

               Open          High           Low         Close     Adj Close  \
count  66008.000000  66008.000000  66008.000000  66008.000000  66008.000000   
mean       0.986197      0.991757      0.524655      0.529674      0.529674   
min       -0.108000     -0.100000     -0.235000     -0.105000     -0.105000   
25%        0.040000      0.040000      0.035000      0.038000      0.038000   
50%        0.105000      0.105000      0.100000      0.100000      0.100000   
75%        0.900000      0.915000      0.890000      0.905000      0.905000   
max     3000.002930   3000.002930      2.403000      2.475000      2.475000   
std       36.923778     36.923755      0.746848      0.748044      0.748044   

        Volume  Tbill yield filled            Nt            rf  
count  66008.0        66008.000000  66008.000000  66008.000000  
mean       0.0            0.529674      1.020664      0.000015  
min        0.0           -0.105000      1.000000     -0.000003  
25%        0.0            0.038000      1.000000      0.000001  
50%        0.0            0.100000      1.000000      0.000003  
75%        0.0            0.905000      1.000000      0.000025  
max        0.0            2.475000      3.000000      0.000130  
std        0.0            0.748044      0.147282      0.000022  

[8 rows x 28 columns]

In [310]:
# calculate excess returns
crsp['ExcessRets'] = crsp['DailyRets'] - crsp['rf']

In [311]:
# calculate annualized_return
annualized_return = (1 + crsp['ExcessRets'].mean())**252 - 1
annualized_return

-0.7774552194719779

In [312]:
# calculate std
annualized_std_dev = crsp['ExcessRets'].std() * (252**0.5)
annualized_std_dev

0.4198221502660796

In [316]:
# find sharpe ratio
sharpe_ratio = (annualized_return - crsp['rf']*365) / annualized_std_dev
sharpe_ratio.mean()

-1.8648299339503336

Conclusion: the strategy is not working